components/brave_extension/extension/brave_extension/background/reducers/shieldsPanelReducer.ts

```diff
@@ -382,12 +382,20 @@ export default function shieldsPanelReducer (
         console.error('Active tab not found')
         break
       }
-      state = shieldsPanelState.saveCosmeticFilterRuleExceptions(state, action.tabId, action.exceptions)
+      if (action.frameId === 0) {
+        // Non-scriptlet cosmetic filters are only applied on the top-level frame
+        state = shieldsPanelState.saveCosmeticFilterRuleExceptions(state, action.tabId, action.exceptions)
+        chrome.tabs.sendMessage(action.tabId, {
+          type: 'cosmeticFilteringBackgroundReady',
+          hide1pContent: tabData.firstPartyCosmeticFiltering,
+          generichide: action.generichide
+        })
+      }
       chrome.tabs.sendMessage(action.tabId, {
-        type: 'cosmeticFilteringBackgroundReady',
-        scriptlet: action.scriptlet,
-        hide1pContent: tabData.firstPartyCosmeticFiltering,
-        generichide: action.generichide
+        type: 'cosmeticFilteringInjectScriptlet',
```

Comment:
So now we're passing two messages in a row for every top level frame? Could we do a combined message to avoid the overhead, or am I missing something? Something like

```
        chrome.tabs.sendMessage(action.tabId, {
          type: 'cosmeticFilteringReady',
          hide1pContent: tabData.firstPartyCosmeticFiltering,
          generichide: action.generichide,
          scriptlet: action.frameId !== 0 ? action.scriptlet : undefined
        }, { frameId: action.frameId })
```

and in the handler

Reply:
It was originally combined into the same message, I figured splitting them provides a better separation of concerns since part of it is only sent for top-level frames. If message passing is particularly expensive I can combine them again.